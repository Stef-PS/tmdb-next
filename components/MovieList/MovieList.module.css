.movieList {
  display: grid;
  grid-auto-flow: row;
  grid-template-columns: 1fr 1fr;
  gap: var(--spacing-xxs);
}
