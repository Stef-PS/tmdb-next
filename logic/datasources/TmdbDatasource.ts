import axios from 'axios'

const LOCALE = 'en-US'
const API_URL = 'https://api.themoviedb.org/3'

export interface SearchMovie {
  id: number,
  poster_path: string,
  adult: boolean,
  overview: string,
  release_date: string,
  genre_ids: number[],
  original_title: string,
  original_language: string,
  title: string,
  backdrop_path: string,
  popularity: number,
  vote_count: number,
  video: boolean,
  vote_average: number,
}

export interface SearchMovieResult {
  page: number,
  results: SearchMovie[],
  total_results: number,
  total_pages: number,
}

export interface Configuration {
  images: {
    baseUrl: string,
    secureBaseUrl: string,
    posterSizes: string[]
  }
}

export class TmdbDatasource {
  private apiKey = process.env.TMDB_API_KEY ?? 'test'
  private options = {
    timeout: 2500,
    headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' }
  }

  searchMovie(query: string, page = 1): Promise<any> {
    const url = `${API_URL}/search/movie?api_key=${this.apiKey}&query=${query}&page=${page}&language=${LOCALE}`
    return axios.get(url, this.options)
      .then((response) => response.data)
      .catch(error => {
        return Promise.reject(error)
      })
  }

  getConfiguration(): Promise<Configuration> {
    const url = `${API_URL}/configuration?api_key=${this.apiKey}`
    return axios.get(url, this.options)
      .then((response) => {
        const {
          base_url: baseUrl,
          secure_base_url: secureBaseUrl,
          poster_sizes: posterSizes
        } = response.data.images
        return { images: { baseUrl, secureBaseUrl, posterSizes } }
      })
      .catch(error => {
        return Promise.reject(error)
      })
  }
}
