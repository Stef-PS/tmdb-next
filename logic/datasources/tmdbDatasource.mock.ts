export const searchMovieResultMock = (query: string, page: number, pageSize: number, total: number) => {
  const results = []
  for (let i = pageSize * (page - 1); i < Math.min(pageSize * page, total); i++) {
    results.push({
      id: i,
      poster_path: `poster_path_${i}`,
      adult: false,
      overview: `overview_${i}`,
      release_date: 'release_date',
      genre_ids: [1, 2],
      original_title: `original_title_${i}`,
      original_language: 'original_language',
      title: `title_${query}_${i}`,
      backdrop_path: `backdrop_path_${i}`,
      popularity: 1,
      vote_count: 1,
      video: false,
      vote_average: 1,
    })
  }
  return {
    page: page,
    results,
    total_results: total,
    total_pages: page,
  }
}

export const getConfigurationMock = () => {
  return {
    images: {
      base_url: 'http://image.tmdb.org',
      secure_base_url: 'https://image.tmdb.org',
      poster_sizes: ['w500']
    }
  }
}

export const getMovieMock = (id: number) => ({
  adult: false,
  backdrop_path: `backdrop_path_${id}`,
  belongs_to_collection: {
    id: id,
    name: `name_${id}`,
    poster_path: `poster_path_${id}`,
    backdrop_path: `backdrop_path_${id}`
  },
  budget: id,
  genres: [{ id: id, name: `name_${id}` }],
  homepage: `homepage_${id}`,
  id: id,
  imdb_id: `imdb_id_${id}`,
  original_language: 'original_language',
  original_title: `original_title_${id}`,
  overview: `overview_${id}`,
  popularity: id,
  poster_path: `poster_path_${id}`,
  production_companies: [{ id: id, logo_path: `logo_path_${id}`, name: `name_${id}`, origin_country: `origin_country_${id}` }],
  production_countries: [{ iso_3166_1: `iso_3166_1_${id}`, name: `name_${id}` }],
  release_date: 'release_date',
  revenue: id,
  runtime: id,
  spoken_languages: [{ english_name: `english_name_${id}`, iso_639_1: `iso_639_1_${id}`, name: `name_${id}` }],
  status: 'status',
  tagline: `tagline_${id}`,
  title: `title_${id}`,
  video: false,
  vote_average: id,
  vote_count: id
})
