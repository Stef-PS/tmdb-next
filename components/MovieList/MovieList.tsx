import MovieCard from '../MovieCard'
import { Configuration, SearchMovie } from "../../logic/datasources/TmdbDatasource"
import styles from './MovieList.module.css'

interface MovieListProps {
  movies: SearchMovie[],
  config: Configuration
}

const MovieList = ({ movies, config }: MovieListProps) => {
  return (
    <ul className={styles.movieList}>
      {movies.map((movie: SearchMovie) => <MovieCard key={movie.id} movie={movie} config={config} />)}
    </ul>
  )
}

export default MovieList
