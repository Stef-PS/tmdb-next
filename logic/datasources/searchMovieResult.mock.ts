import { SearchMovie, SearchMovieResult } from './TmdbDatasource'

export const searchMovieResultMock = (query: string, page: number, pageSize: number, total: number): SearchMovieResult => {
  const results: SearchMovie[] = []
  for (let i = pageSize * (page - 1); i < Math.min(pageSize * page, total); i++) {
    results.push({
      id: i,
      poster_path: `poster_path_${i}`,
      adult: false,
      overview: `overview_${i}`,
      release_date: 'release_date',
      genre_ids: [1, 2],
      original_title: `original_title_${i}`,
      original_language: 'original_language',
      title: `title_${query}_${i}`,
      backdrop_path: `backdrop_path_${i}`,
      popularity: 1,
      vote_count: 1,
      video: false,
      vote_average: 1,
    })
  }
  return {
    page: page,
    results,
    total_results: total,
    total_pages: page,
  }
}
