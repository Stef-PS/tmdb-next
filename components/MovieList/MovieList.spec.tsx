import { render } from '@testing-library/react'
import { Configuration, SearchMovie } from '../../logic/datasources/TmdbDatasource'
import '@testing-library/jest-dom'
import MovieList from './MovieList'

const movies = [
  { title: 'movie title 1', id: 1, releaseDate: '2001-01-01', voteAverage: 2, posterPath: 'posterPath1' },
  { title: 'movie title 2', id: 2, releaseDate: '2002-01-01', voteAverage: 4, posterPath: 'posterPath2' },
  { title: 'movie title 3', id: 3, releaseDate: '2003-01-01', voteAverage: 6, posterPath: 'posterPath3' },
] as SearchMovie[]

const config = {
  images: {
    baseUrl: 'https://image.tmdb.org/t/p/',
    secureBaseUrl: 'https://image.tmdb.org/t/p/',
    posterSizes: ['w500']
  }
} as Configuration


describe('Movie List', () => {
  it('should render all titles', () => {
    const { getByText } = render(<MovieList movies={movies} config={config} />)
    movies.forEach(movie => {
      expect(getByText(movie.title)).toBeInTheDocument()
    })
  })

  it('should render a list of images', () => {
    const { getByAltText } = render(<MovieList movies={movies} config={config} />)
    movies.forEach(movie => {
      expect(getByAltText(movie.title)).toBeInTheDocument()
    })
  })

  it('should render the right number of stars', () => {
    const { container } = render(<MovieList movies={movies} config={config} />)
    expect(container.querySelectorAll('.star')?.length).toEqual(6)
  })

  it('should render a list of release years', () => {
    const { getByText } = render(<MovieList movies={movies} config={config} />)
    movies.forEach(movie => {
      expect(getByText(movie.releaseDate.substring(0,4))).toBeInTheDocument()
    })
  })

  it('should render an empty list', () => {
    const { container } = render(<MovieList movies={[]} config={config} />)
    expect(container.querySelectorAll('li')?.length).toEqual(0)
  })

  it('should render the right number of li tags', () => {
    const { container } = render(<MovieList movies={movies} config={config} />)
    expect(container.querySelectorAll('li')?.length).toEqual(3)
  })

  it('should render the movie list unchanged', () => {
    const { container } = render(<MovieList movies={movies} config={config} />)
    expect(container.innerHTML).toMatchSnapshot()
  })
})
