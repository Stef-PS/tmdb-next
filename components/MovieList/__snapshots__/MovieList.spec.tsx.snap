// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie List should render the movie list unchanged 1`] = `"<ul class=\\"movieList\\"><li class=\\"card\\"><div class=\\"poster\\"><span style=\\"box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;\\"><span style=\\"box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 150% 0px 0px 0px;\\"></span><img alt=\\"movie title 1\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7\\" decoding=\\"async\\" data-nimg=\\"responsive\\" style=\\"position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;\\"><noscript></noscript></span></div><div class=\\"info\\"><div class=\\"title\\">movie title 1</div><div class=\\"aside\\"><div class=\\"year\\">2001</div><div class=\\"rating\\"><span class=\\"star\\">★</span><span class=\\"starEmpty\\">★</span><span class=\\"starEmpty\\">★</span><span class=\\"starEmpty\\">★</span><span class=\\"starEmpty\\">★</span></div></div></div></li><li class=\\"card\\"><div class=\\"poster\\"><span style=\\"box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;\\"><span style=\\"box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 150% 0px 0px 0px;\\"></span><img alt=\\"movie title 2\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7\\" decoding=\\"async\\" data-nimg=\\"responsive\\" style=\\"position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;\\"><noscript></noscript></span></div><div class=\\"info\\"><div class=\\"title\\">movie title 2</div><div class=\\"aside\\"><div class=\\"year\\">2002</div><div class=\\"rating\\"><span class=\\"star\\">★</span><span class=\\"star\\">★</span><span class=\\"starEmpty\\">★</span><span class=\\"starEmpty\\">★</span><span class=\\"starEmpty\\">★</span></div></div></div></li><li class=\\"card\\"><div class=\\"poster\\"><span style=\\"box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;\\"><span style=\\"box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 150% 0px 0px 0px;\\"></span><img alt=\\"movie title 3\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7\\" decoding=\\"async\\" data-nimg=\\"responsive\\" style=\\"position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;\\"><noscript></noscript></span></div><div class=\\"info\\"><div class=\\"title\\">movie title 3</div><div class=\\"aside\\"><div class=\\"year\\">2003</div><div class=\\"rating\\"><span class=\\"star\\">★</span><span class=\\"star\\">★</span><span class=\\"star\\">★</span><span class=\\"starEmpty\\">★</span><span class=\\"starEmpty\\">★</span></div></div></div></li></ul>"`;
